### PR TITLE
Add a convention to replace shadow foreign keys with non-shadow ones.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Properties are be matched using the following patterns:
1. Navigation to principal name + corresponding principal key property name
2. Principal type name + corresponding principal key property name
3. Corresponding principal key property name
4. PK on the dependent entity for unique FKs

Add a way to specify non-primary key to InternalEntityBuilder.

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
-﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs]: Why is this an extension method?

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs [hunk 2]
```diff
@@ -76,11 +75,40 @@ private InternalKeyBuilder Key(IReadOnlyList<Property> properties, Configuration
...
+            if (properties == null)
+            {
+                return null;
+            }
+
+            return _keyBuilders.GetOrAdd(
+                () => Metadata.TryGetKey(properties),
+                () => Metadata.AddKey(properties),
+                key => new InternalKeyBuilder(key, ModelBuilder),
+                null,
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: named arg?

File: src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs [hunk 3]
```diff
@@ -12,16 +12,19 @@ public class ConventionsDispatcher
     {
         public ConventionsDispatcher()
         {
-            EntityTypeConventions = new List<IEntityTypeConvention>();
+            OnEntityTypeAddedConventions = new List<IEntityTypeConvention>();
+            OnForeignKeyAddedConventions = new List<IRelationshipConvention>();
         }
 
-        public virtual IList<IEntityTypeConvention> EntityTypeConventions { get; }
+        public virtual IList<IEntityTypeConvention> OnEntityTypeAddedConventions { get; }
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs]: Consider a name that doesn't read like a hook method.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs [hunk 4]
```diff
@@ -706,49 +713,211 @@ public virtual InternalRelationshipBuilder Relationship([NotNull] ForeignKey for
...
+                // Create foreign key properties in shadow state if no properties are specified
+                var baseName = (string.IsNullOrEmpty(navigationToPrincipal) ? principalType.SimpleName : navigationToPrincipal);
+
+                var fkProperties = new List<Property>();
+                foreach (var keyProperty in principalKey.Properties)
+                {
+                    var index = -1;
+                    while (true)
+                    {
+                        var name = baseName + keyProperty.Name + (++index > 0 ? index.ToString() : "");
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: string.Empty?

**Comment by anpete** [comment]
:shipit:

File: src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs (hunk 1 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs]: Will move to EntityType

File: src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs]: I'm open to suggestions. It represents a list of entity type conventions that run when a new entity type is added.
Note that in the future there might be also entity type conventions that run when something different happens.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs (hunk 4 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: We've discussed this with Arthur and "" is as readable in most cases and uses less characters.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: I'll add it, but you won't like it. It's onNewKeyAdded:

Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: Is it a delegate? :-)

File: src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs (hunk 3 above)
Comment (anpete) [inline src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs]: EntityTypeAddedConventions?

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: it is!

File: src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/ModelConventions/ConventionsDispatcher.cs]: ok

**Pull request merged**